### PR TITLE
MW-1158: Add page for Program Selection for Issue mobile

### PR DESCRIPTION
**Pull request by sniedzielski**
TICKET: https://openlmis.atlassian.net/browse/MW-1158

New feature:
- 'ProgramSelection' page for issue - generic approach (only 'reducers' and 'routes' are different in comparison with 'Adjustment')
- keep saved 'Adjustment' when user want to switch into 'Issue' page

Changes:
- BlockList changes - reason in header, 'quantity' instead of 'SoH', removed occured date

### CODE REVIEW

File: src/stock-adjustment-mobile/add-products-page/add-product-page.jsx [hunk 1]
```diff
@@ -214,8 +215,8 @@ const AddProductsPage = ({}) => {
                                                 />
                                                 {renderLotSelect(name, values.items[index].product, values.items[index])}
                                                 <ReadOnlyField
-                                                    name="expiryDate"
-                                                    label="Expiry Date"
+                                                    name="occuredDate"
```
Comment (rstencel) [inline src/stock-adjustment-mobile/add-products-page/add-product-page.jsx:218]: `occured...` -> `occurred...`

File: src/stock-adjustment-mobile/components/block-list.component.jsx [hunk 2]
```diff
@@ -15,21 +15,29 @@
...
-                                    if (headerToDisplay === display.key) {
-                                        return ( 
-                                            <div className="react-block-list-container-header-text" style={{height: "16px", display: "flex", marginBottom: "16px", marginTop: "16px"}}>
-                                                <h3 style={{marginLeft: "5%", fontWeight: "bold", fontSize: "18px"}}>{row[display.key]}</h3>
-                                            </div>
-                                        );  
+                                    if (headersToDisplay.includes(display.key)) { 
+                                        if (headersCounter < headersToDisplay.length - 1) {
+                                            headerText += row[display.key] + " ";
+                                            headersCounter += 1;
```
Comment (rstencel) [inline src/stock-adjustment-mobile/components/block-list.component.jsx:32]: As far as I understand you are returning value ONLY for the last header to display. When using `map` function a value should be returned from every callback: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/map#when_not_to_use_map

Maybe you could iterate trough the headers outside of `return()` function, use `reduce` or `forEach` function to aggregate the string and display it only for the last header.

File: src/stock-adjustment-mobile/edit-product-page/edit-product-page.jsx [hunk 3]
```diff
@@ -263,8 +264,8 @@ const EditProductPage = ({ offlineService }) => {
                                                 />
                                                 {renderLotSelect(name, values.items[index].product, values.items[index])}
                                                 <ReadOnlyField
-                                                    name="expiryDate"
-                                                    label="Expiry Date"
+                                                    name="occuredDate"
```
Comment (rstencel) [inline src/stock-adjustment-mobile/edit-product-page/edit-product-page.jsx:267]: `occured...` -> `occurred...`

File: src/stock-issue-mobile/issue-app.jsx [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+import { useDispatch, useSelector } from 'react-redux';
+import { setUserHomeFacility } from '../stock-adjustment-mobile/reducers/facilities';
+import ProgramSelect from '../stock-adjustment-mobile/program-select';
+
+const IssueApp = ({
+        facilityFactory,
+        existingStockOrderableGroupsFactory,
+        stockReasonsFactory,
+        offlineService,
+    }) => {
```
Comment (rstencel) [inline src/stock-issue-mobile/issue-app.jsx:27]: Too much indentation:
```
const IssueApp = ({
    facilityFactory,
    existingStockOrderableGroupsFactory,
    stockReasonsFactory,
    offlineService,
}) => {
```

File: src/stock-issue-mobile/issue-app.jsx [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+    const dispatch = useDispatch();
+    const userHomeFacility = useSelector(state => state.facilities.userHomeFacility);
+
+    useEffect(
+        () => {
+            facilityFactory.getUserHomeFacility()
+                .then(facility => dispatch(setUserHomeFacility(facility)))
+        },
+        [facilityFactory]
+    );
```
Comment (rstencel) [inline src/stock-issue-mobile/issue-app.jsx:38]: When only one line is in the body you can use `() => (...` or even `() => ...`:
```
useEffect(() => (
  facilityFactory.getUserHomeFacility().then(facility => dispatch(setUserHomeFacility(facility)))
), [facilityFactory]);
```
or longer
```
useEffect(() => facilityFactory.getUserHomeFacility().then(facility => dispatch(setUserHomeFacility(facility))), [facilityFactory]);
```

File: src/stock-issue-mobile/issue-app.jsx [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+
+    const menu = document.getElementsByClassName("header ng-scope")[0];
+
+    useEffect(() => {
+        menu.style.display = "";
+    }, [menu]);
```
Comment (rstencel) [inline src/stock-issue-mobile/issue-app.jsx:44]: Same as above:
```
useEffect(() => menu.style.display = "", [menu]);
```

File: src/stock-issue-mobile/reducers/adjustment.jsx [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+        },
+        resetAdjustment: (state, action) => {
+            state.adjustment = []
+        }
+    }
+});
+
+export const {setAdjustment} = adjustmentSlice.actions;
+export const {appendToAdjustment} = adjustmentSlice.actions;
+export const {resetAdjustment} = adjustmentSlice.actions;
```
Comment (rstencel) [inline src/stock-issue-mobile/reducers/adjustment.jsx:38]: I think you can simplify to: 
```
export const { setAdjustment, appendToAdjustment, resetAdjustment } = adjustmentSlice.actions;
```

File: src/stock-issue-mobile/reducers/adjustment.jsx [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+        adjustment: []
+    },
+    reducers: {
+        setAdjustment: (state, action) => {
+            state.adjustment = action.payload;
+        },
+        appendToAdjustment: (state, action) => {
+            state.adjustment.push(action.payload);
+        },
+        resetAdjustment: (state, action) => {
```
Comment (rstencel) [inline src/stock-issue-mobile/reducers/adjustment.jsx:30]: As you are not using `action` I think you can change to:
```
resetAdjustment: (state, _) => {
```
or
```
resetAdjustment: (state) => {
```

File: src/stock-issue-mobile/reducers/program.jsx [hunk 9]
```diff
@@ -0,0 +1,36 @@
...
+export const programSlice = createSlice({
+    name: 'program',
+    initialState: {
+        program: []
+    },
+    reducers: {
+        setProgram: (state, action) => {
+            state.program = action.payload;
+        },
+        resetProgram: (state, action) => {
```
Comment (rstencel) [inline src/stock-issue-mobile/reducers/program.jsx:27]: As you are not using `action` I think you can change to:
```
resetAdjustment: (state, _) => {
```
or
```
resetAdjustment: (state) => {
```

File: src/stock-issue-mobile/reducers/program.jsx [hunk 10]
```diff
@@ -0,0 +1,36 @@
...
+            state.program = null;
+        }
+    }
+});
+
+export const {setProgram} = programSlice.actions;
+export const {resetProgram} = programSlice.actions;
```
Comment (rstencel) [inline src/stock-issue-mobile/reducers/program.jsx:34]: Same as above - I think you can simplify to:
```
export const { setProgram, resetProgram } = programSlice.actions;
```

File: src/stock-issue-mobile/store.jsx [hunk 11]
```diff
@@ -0,0 +1,64 @@
...
+    },
+    middleware: (getDefaultMiddleware) => getDefaultMiddleware({
+        serializableCheck: false
+    }),
+    preloadedState: persistedStore
+});
+
+store.subscribe(() => {
+    saveToLocalStorage(store.getState());
+});
```
Comment (rstencel) [inline src/stock-issue-mobile/store.jsx:62]: If you want you can shorten to:
```
store.subscribe(() => saveToLocalStorage(store.getState()));
```

**Review by rstencel — CHANGES_REQUESTED** [review]
Only one critical issue found regarding use of `map` function. The rest of remarks can be addressed, but don't have to be.

**Comment by sniedzielski** [comment]
@rstencel thanks for review! As you wrote the most important is to fix this issue with 'map' function in BlockList.

File: src/stock-adjustment-mobile/components/block-list.component.jsx (hunk 2 above)
Comment (sniedzielski) [reply src/stock-adjustment-mobile/components/block-list.component.jsx:32]: yes you are right that this approach was bad. Here I commited https://github.com/OpenLMIS/openlmis-stockmanagement-ui/pull/24/commits/f4306c9bcf34f8fde8362e3081adb9630a7ca8f8 this change. Will be merged into this PR.

**Pull request merged**